rocket-chatter-ingestion-server/src/process/prepare/codebase.ts

```diff
@@ -0,0 +1,210 @@
+	private _path: string
+	private _dataDirPath: string
+	private _dataDirName: string
+	private _embeddingsDirPath: string
+	private _embeddingsDirName: string
+
+	private _batchSize: number
+	private _fileProcessor: IFileProcessor
+
+	private _files: ISourceFile[] = []
+	private _batches: number[][] = []
+
+	constructor(path: string, fileProcessor: IFileProcessor, batchSize = 50) {
+		// If path ends with any number of /, remove them
+		if (path.endsWith("/")) path = path.replace(/\/+$/, "")
+
+		this._path = path
+		this._dataDirName = ""
+		this._dataDirPath = ""
+		this._embeddingsDirName = ""
+		this._embeddingsDirPath = ""
+
+		this._batchSize = batchSize
+		this._fileProcessor = fileProcessor
+
+		this.initializeDataDirectory()
+		this.prepareFilesMetadata()
+		this.makeFilesBatches()
+	}
+
+	get embeddingsDirPath(): string {
+		return this._embeddingsDirPath
+	}
+
+	private initializeDataDirectory(removeExisting = true): void {
+		this._dataDirName = `data`
+		this._dataDirPath = path.resolve(this._path, this._dataDirName)
+
+		this._embeddingsDirName = `${this._dataDirName}/embeddings`
+		this._embeddingsDirPath = path.resolve(this._path, this._embeddingsDirName)
+
+		/* Handle data directory */
+		if (removeExisting && existsSync(this._dataDirPath))
+			rmSync(this._dataDirPath, { recursive: true })
+		mkdirSync(this._dataDirPath)
+
+		/* Handle embeddings directory */
+		mkdirSync(this._embeddingsDirPath)
+	}
+
+	private prepareFilesMetadata() {
+		const extensions = ["ts", "tsx", "js", "jsx"]
+
+		console.log(`🕒 Preparing metadata for files: *.${extensions.join(", *.")}`)
+		{
+			const globPatterns = extensions.map((x) => `**/*.${x}`)
+			for (const pattern of globPatterns) {
+				const files = glob
+					.sync(`${this._path}/${pattern}`)
+					.map((x) => new SourceFile(x))
+				this._files.push(...files)
+			}
+		}
+		console.log(`✅ Prepared metadata for ${this._files.length} files\n`)
+	}
+
+	private makeFilesBatches() {
+		const batches: number[][] = []
+		for (let i = 0; i < this._files.length; i += this._batchSize)
+			batches.push([i, i + this._batchSize])
+
+		this._batches = batches
+	}
+
+	private writeNodesToFile(nodes: Record<string, DBNode>, fileName: string) {
+		const entries = Object.entries(nodes)
+		if (entries.length === 0) return 0
+		const batch = Object.fromEntries(entries)
+		writeFileSync(
+			path.resolve(this._dataDirPath, fileName),
+			JSON.stringify(batch, null, 2)
+		)
+
+		return entries.length
+	}
+
+	private async processFilesBatch(
+		batchNumber: number,
+		start: number,
+		end: number
+	): Promise<number> {
+		let nNodesProcessed = 0
+
+		console.log(`🕒 Processing ${start}-${end} files`)
+		{
+			let nodes: Record<string, DBNode> = {}
+
+			/* Step 1 */
+			// try {
+			const files = this._files.slice(start, end)
+			const jobs = files.map((x) => this._fileProcessor.process(x, nodes))
+			await Promise.all(jobs)
+			// } catch {
+			// 	console.error(`Error in processing ${start}-${end} files`)
```

Comment:
also remove unused code here

Reply:
Actually this is meant to be included but left commented by accident. Now including it.